Browse: Return additional info

Hwesta opened the pull request:
Return more information in a 'properties' entry in the browse dict about the browse entries. Currently, returns number of objects in a directory and (for local filesystems) file size.

refs #7853

mistydemeo commented on storage_service/locations/models/space.py:
```diff
@@ -143,16 +143,31 @@ def get_child_space(self):
+
+        'entries' is a list of strings, one for each entry in that directory, both file-like and folder-like.
+        'directories' is a list of strings for each folder-like entry. Each entry should also be listed in 'entries'.
+        'properties' is a dictionary that may contain additional information for the entries.  Keys are the entry name found in 'entries', values are a dictionary containing extra information. 'properties' may not contain all values from 'entries'.
```
I know it's provided in the sample below, but mind explicitly documenting what properties currently exist and when they're reported?

Hwesta replied:
I had anticipated 'properties' being more free-form than that, since not all Spaces will provide all properties (eg Duracloud doesn't do file size though LocalFS does, so I'm hesitant to provide a definitive list. The consuming code should handle if the data isn't there.  Do you still think it'd be helpful to provide a partial list with caveats?

mistydemeo replied:
Sure, that'd help! I'm looking at this as a consumer of the data later, wanting to know what kind of data is actually going to be available - otherwise I wouldn't be sure what I could do with it.

If we could document in the spaces what properties they can provide that would also be helpful!

Hwesta replied:
Added info about the properties that are/can be returned.

mistydemeo commented on storage_service/locations/models/space.py:
```diff
@@ -351,6 +366,15 @@ def _create_local_directory(self, path, mode=None):
+        """
+        Returns all the files in a directory, including children.
+        """
+        total_files = 0
```
Sorry to golf, but could reduce this to a one-liner via `reduce`:

``` python
return reduce(lambda count, (dir, dirs, files): count + len(files), os.walk(path), 0)
```

mistydemeo commented:
:+1:

mistydemeo commented:
Though, this is making me think... I wonder if it would make sense to send these transfer components as partial, or complete, bags? Bags contain a metadata file that lets you know the expected contents (numbers of files/directories, expected sizes), which are the same properties we're sending as a part of the JSON response here.

Hwesta committed:
Browse: Allow returning additional info

refs #7853

Return more information in a 'properties' entry in the browse dict about
the browse entries.  Currently empty.

Hwesta committed:
Browse: Return directory object count, object size

Return extra info in properties dict.

Hwesta commented:
Hmm, maybe. On the other hand, we may want to send other things not stored in a bag back, like the last modified timestamp. Automation tools will use that to only run new or changed transfers through.

mistydemeo commented:
If we need extra properties that bags don't provide, then yeah, makes sense not to use that.
